### ISSUE TITLE
Content-Length added to HTTP 304 responses not compliant with RFC7230

### ISSUE THREAD

**Issue by edmorley** [ninenines/cowboy#1153]
[RFC7230 section 3.3.2](https://tools.ietf.org/html/rfc7230#section-3.3.2) says:
> A server MAY send a Content-Length header field in a 304 (Not Modified) response to a conditional GET request (Section 4.1 of [RFC7232]); a server MUST NOT send Content-Length in such a response unless its field-value equals the decimal number of octets that would have been sent in the payload body of a 200 (OK) response to the same request.

ie: HTTP 304 responses (which must not contain a body) can either set a `Content-Length` of what the body *would* have been, or must not set it at all. (And in fact, [RFC7232 section 4.1](https://tools.ietf.org/html/rfc7232#section-4.1) suggests that it's best to omit the header entirely to reduce transfer size.)

Currently if an app:
* asks cowboy to respond with a 304
* doesn't pass an explicit Content-Length (which is allowed)
* and correctly doesn't pass a body to cowboy (since HTTP 304s shouldn't have a body)

...then cowboy adds `Content-Length: 0` to the response, which is in violation of RFC7230. 

In this case cowboy has no way of knowing what the original body size was (since the app truncated it already), so should just not add a Content-Length header if one wasn't already set by the app.

**STR:**
1) Clone this repo and checkout revision c2beff7 
2) Install erlang 18.3
3) `cd examples/hello_world/`
4) Modify `toppage_handler.erl` to match the below
5) `make run`
6) In another terminal, `curl -i localhost:8080`

**Expected:**
```
HTTP/1.1 304 Not Modified
date: Wed, 12 Apr 2017 21:05:03 GMT
server: Cowboy
```

**Actual:**
```
HTTP/1.1 304 Not Modified
content-length: 0
date: Wed, 12 Apr 2017 21:05:03 GMT
server: Cowboy
```

**toppage_handler.erl:**
```erlang
-module(toppage_handler).

-export([init/2]).

init(Req0, Opts) ->
	Req = cowboy_req:reply(304, Req0),
	{ok, Req, Opts}.
```

**Comment by edmorley** [comment]
Similarly for HTTP 204s, cowboy adds a `Content-Type: 0` when [RFC7230 section 3.3.2](https://tools.ietf.org/html/rfc7230#section-3.3.2) says one should never be set at all:

> A server MUST NOT send a Content-Length header field in any response with a status code of 1xx (Informational) or 204 (No Content).  A server MUST NOT send a Content-Length header field in any 2xx (Successful) response to a CONNECT request (Section 4.3.6 of [RFC7231]).

STR is the the same as the HTTP 304 STR above, but `toppage_handler.erl` should be:
```erlang
-module(toppage_handler).

-export([init/2]).

init(Req0, Opts) ->
	Req = cowboy_req:reply(204, Req0),
	{ok, Req, Opts}.
```

**Expected:**
```
HTTP/1.1 204 No Content
date: Thu, 13 Apr 2017 15:17:43 GMT
server: Cowboy
```

**Actual:**
```
HTTP/1.1 204 No Content
content-length: 0
date: Thu, 13 Apr 2017 15:17:43 GMT
server: Cowboy
```

**Comment by essen** [comment]
Yes it seems to be missing some cases when replying, thanks. Patch welcome.

**Comment by essen** [comment]
This has been fixed for 204. The problem still exists for 304. I will get on to rfc7232 soon.

**Comment by edmorley** [comment]
Many thanks :-)

**Comment by edmorley** [comment]
(For posterity) Fixed by:
https://github.com/ninenines/cowboy/commit/b0519af42a4d3ee7e345967206bd1d0050e9061b
https://github.com/ninenines/cowboy/commit/292039362a6125dfd0a163d5b0a49b800bf80b11
